src/auto_note.cpp

```diff
@@ -0,0 +1,454 @@
+    }
+}
+
+void auto_note_settings::default_initialize()
+{
+    // Perform clear for good measure
+    this->clear();
+
+    // Fetch all know map extra types and iterate over them
+    for( auto extra : MapExtras::mapExtraFactory().get_all() ) {
+        if( extra.autonote ) {
+            this->autoNoteEnabled.insert( extra.id.str() );
+        }
+    }
+}
+
+void auto_note_settings::set_discovered( const string_id<map_extra> &mapExtId )
+{
+    this->discovered.insert( mapExtId.str() );
+}
+
+bool auto_note_settings::was_discovered( const string_id<map_extra> &mapExtId ) const
+{
+    return this->discovered.count( mapExtId.str() ) != 0;
+}
+
+void auto_note_settings::show_gui()
+{
+    // Create local GUI instance and show
+    auto_note_manager_gui gui{ };
+    gui.show();
+
+    // If settings were changed, we need to do a save
+    if( gui.was_changed() ) {
+        this->save();
+    }
+}
+
+bool auto_note_settings::has_auto_note_enabled( const std::string &mapExtId ) const
+{
+    return this->autoNoteEnabled.count( mapExtId ) != 0;
+}
+
+void auto_note_settings::set_auto_note_status( const std::string &mapExtId, const bool enabled )
+{
+    if( enabled ) {
+        this->autoNoteEnabled.insert( mapExtId );
+    } else if( this->has_auto_note_enabled( mapExtId ) ) {
+        this->autoNoteEnabled.erase( mapExtId );
+    }
+}
+
+bool auto_note_settings::has_auto_note_enabled( const string_id<map_extra> &mapExtId ) const
+{
+    return this->has_auto_note_enabled( mapExtId.str() );
+}
+
+void auto_note_settings::set_auto_note_status( const string_id<map_extra> &mapExtId,
+        const bool enabled )
+{
+    this->set_auto_note_status( mapExtId.str(), enabled );
+}
+
+void auto_note_manager_gui::initialize()
+{
+    // Clear all data
+    this->mapExtraCache.clear();
+
+    // Fetch reference to current auto note settings
+    const auto_note_settings &settings = get_auto_notes_settings();
+
+    // Fetch all know map extra types and iterate over them
+    for( auto extra : MapExtras::mapExtraFactory().get_all() ) {
+        // Ignore all extras that have autonote disabled in the JSON.
+        // This filters out lots of extras users shouldnt see (like "normal")
+        if( !extra.autonote ) {
+            continue;
+        }
+
+        // Check if auto note is enabled for this map extra type
+        bool isAutoNoteEnabled = settings.has_auto_note_enabled( extra.id );
+
+        // Insert into cache
+        this->mapExtraCache.emplace( std::make_pair( extra.id.str(), std::make_pair( extra,
+                                     isAutoNoteEnabled ) ) );
+
+        // Only insert into display cache if map extra was already discovered. This
+        // is done in order to avoid spoilers
+        if( settings.was_discovered( extra.id ) ) {
+            this->displayCache.push_back( extra.id.str() );
+        }
+    }
+}
+
+auto_note_manager_gui::auto_note_manager_gui()
+{
+    // Initialize GUI state
+    this->initialize();
```

Comment:
Why doesn't the constructor do this initialization directly? Why do you need a separate function (that is **only** called from here) for that?